client/allocrunner/taskrunner/task_runner.go

```diff
@@ -1135,6 +1135,22 @@ func (tr *TaskRunner) buildTaskConfig() *drivers.TaskConfig {
 				Options:  interpolatedNetworks[0].DNS.Options,
 			}
 		}
+
+		// merge in the DNS set by any CNI plugins
+		netStatus := tr.allocHookResources.GetAllocNetworkStatus()
+		if netStatus != nil && netStatus.DNS != nil {
+			if dns == nil {
+				dns = &drivers.DNSConfig{
+					Servers:  netStatus.DNS.Servers,
+					Searches: netStatus.DNS.Searches,
+					Options:  netStatus.DNS.Options,
+				}
+			} else {
+				dns.Servers = append(netStatus.DNS.Servers, dns.Servers...)
```

Comment:
While I understand the goal here, I am not sure if merging DNS is the proper way to go here. For instance if `options` has `rotate` set then there would be no guarantee that the first server is actually used. https://developer.hashicorp.com/nomad/docs/job-specification/network#dns also says that the defaults from the client host are used, so we don't necessarily know  what is set etc… I fear that this might end up being a bit fragile.

Reply:
As you've noted, there are three potential sources for DNS configuration here: the output of CNI plugins, the user-provided DNS config, and whatever is on the client. This code path only involves the first two. The task driver is then responsible for what happens after that... the built-in task drivers use [`GenerateDNSMount`](https://github.com/hashicorp/nomad/blob/v1.7.4/drivers/shared/resolvconf/mount.go#L16) to create the `/etc/resolv.conf`. (So for example, you could have a third-party task driver that requires a particular CNI plugin be used, in which case it'd use that value alone and not get anything from the client host.)

All that being said, any DNS configuration from the CNI plugins is frankly likely to conflict with any jobspec `network.dns` configuration. Our default network configuration for `bridge` networking doesn't generate any DNS values. So if you're intentionally using a CNI plugin that does, you _should_ know not to set conflicting values on the `network.dns` block. Otherwise yes, it's fragile... we want to give users knobs here but its up to them to use them responsibly.

Could definitely do with some warnings around `network.dns` usage in the docs though.

Reply:
I'm having second thoughts on the _order_ though here... lemme get some input from the team on that.

Reply:
So since DNS is usually at fault when there are issues I'd prefer a "stable" configuration like the following:

 * By default use the resolv.conf like now (ie I have a custom DNS set in docker which nicely ends up in the resolv.conf). Now here being as in current master before this patch.
 * If DNS is set by CNI, override the default.
 * if DNS is set in the task, override also whatever was set before. I would consider setting DNS on the task level a rather specific and advanced configuration, people doing that can as well set a DNS server that knows how to handle `.consul`

In a perfect world I'd love to be able to customize the template for the default bridge CNI configuration (with and without tproxy) so I could globally override that for all bridge users without having to go into every task and change mode to `cni/my_cni`.

Reply:
Aye, that's pretty much what the rest of the Nomad team suggested we do here, so I was definitely off the rails a bit :grinning: 

> In a perfect world I'd love to be able to customize the template for the default bridge CNI configuration (with and without tproxy) so I could globally override that for all bridge users without having to go into every task and change mode to cni/my_cni.

Definitely an interesting idea. A little challenging from the perspective of fingerprinting, because if there's behavior the application needs from that specific bridge configuration there's no reasonable way for us to tell the server which clients have that configuration set.

Reply:
Not sure why this would require fingerprinting. In it's easiest implementation one could just specify an alternate template in the client-configuration for https://github.com/hashicorp/nomad/blob/20cfbc82d39e944aa63a22f8170fb8062bc2adcf/client/allocrunner/networking_bridge_linux.go#L152 (we probably need templating to play nicely with subnets etc…).

We could even add cni_consul to that list by default (assuming nomad) ships it or so and have it activate/deactivate based on cni args. This way the same template could be used independent of whether tproxy is used or not.